langchain/src/agents/chat_convo/index.ts

```diff
@@ -0,0 +1,195 @@
+import {
+  Agent,
+  Tool,
+  AgentInput,
+  StaticAgent,
+  staticImplements,
+} from "../index.js";
+import {
+  SystemMessagePromptTemplate,
+  HumanMessagePromptTemplate,
+  ChatPromptTemplate,
+  MessagesPlaceholder,
+} from "../../prompts/index.js";
+import { interpolateFString } from "../../prompts/template.js";
+import {
+  PREFIX,
+  SUFFIX,
+  FORMAT_INSTRUCTIONS,
+  TEMPLATE_TOOL_RESPONSE,
+} from "./prompt.js";
+import { BaseLanguageModel } from "../../base_language/index.js";
+import {
+  AgentStep,
+  BaseChatMessage,
+  AIChatMessage,
+  HumanChatMessage,
+} from "../../schema/index.js";
+import {
+  BaseOutputParser,
+  SerializedOutputParser,
+} from "../../output_parsers/base.js";
+
+export class AgentOutputParser extends BaseOutputParser {
+  parse(text: string): unknown {
+    const cleanedOutput = text.trim();
+    let jsonOutput = cleanedOutput;
+    if (jsonOutput.includes("```json")) {
+      jsonOutput = jsonOutput.split("```json")[1].trimLeft();
+    }
+    if (jsonOutput.startsWith("```")) {
+      jsonOutput = jsonOutput.slice(3).trimLeft();
+    }
+    if (jsonOutput.endsWith("```")) {
+      jsonOutput = jsonOutput.slice(0, -3).trimRight();
+    }
+    const response = JSON.parse(jsonOutput);
+    return { action: response.action, action_input: response.action_input };
+  }
+
+  getFormatInstructions(): string {
+    return FORMAT_INSTRUCTIONS;
+  }
+
+  serialize(): SerializedOutputParser {
+    throw new Error("Method not implemented.");
+  }
+}
+
+export type CreatePromptArgs = {
+  /** String to put after the list of tools. */
+  systemMessage?: string;
+  /** String to put before the list of tools. */
+  humanMessage?: string;
+  /** List of input variables the final prompt will expect. */
+  inputVariables?: string[];
+  /** Output parser to use for formatting. */
+  outputParser?: BaseOutputParser;
+};
+
+type ZeroShotAgentInput = AgentInput;
+
+/**
+ * Agent for the MRKL chain.
+ * @augments Agent
+ * @augments StaticAgent
+ */
+@(staticImplements<StaticAgent>)
+export class ChatConversationalAgent extends Agent {
+  outputParser: BaseOutputParser;
+
+  constructor(input: ZeroShotAgentInput, outputParser: BaseOutputParser) {
+    super(input);
+    this.outputParser = outputParser;
+  }
+
+  _agentType(): string {
+    /** Not turning on serialization until more sure of abstractions. */
+    throw new Error("Method not implemented.");
+  }
+
+  observationPrefix() {
+    return "Observation: ";
+  }
+
+  llmPrefix() {
+    return "Thought:";
+  }
+
+  _stop(): string[] {
+    return ["Observation:"];
+  }
+
+  static validateTools(tools: Tool[]) {
+    const invalidTool = tools.find((tool) => !tool.description);
+    if (invalidTool) {
+      const msg =
+        `Got a tool ${invalidTool.name} without a description.` +
+        ` This agent requires descriptions for all tools.`;
+      throw new Error(msg);
+    }
+  }
+
+  constructScratchPad(steps: AgentStep[]): BaseChatMessage[] {
```

Comment:
do we want to construct the scratchpad as messages for the regular chat agent too?

Reply:
probably not